diff --git a/src/dmenu-4.5/dmenu.c b/src/dmenu-4.5/dmenu.c
index 4ea95f8..cf7f035 100644
--- a/src/dmenu-4.5/dmenu.c
+++ b/src/dmenu-4.5/dmenu.c
@@ -5,6 +5,7 @@
 #include <string.h>
 #include <strings.h>
 #include <unistd.h>
+#include <wordexp.h>
 #include <X11/Xlib.h>
 #include <X11/Xatom.h>
 #include <X11/Xutil.h>
@@ -32,6 +33,7 @@ static void grabkeyboard(void);
 static void insert(const char *str, ssize_t n);
 static void keypress(XKeyEvent *ev);
 static void match(void);
+static void matchfile(char *filestart);
 static size_t nextrune(int inc);
 static void paste(void);
 static void readstdin(void);
@@ -375,6 +377,10 @@ keypress(XKeyEvent *ev) {
 		}
 		break;
 	case XK_Tab:
+		if( strchr(text, ' ')!=NULL ) {
+			matchfile( strchr(text, ' ')+1 );
+			break;
+		}
 		if(!sel)
 			return;
 		strncpy(text, sel->text, sizeof text);
@@ -464,6 +470,35 @@ paste(void) {
 }
 
 void
+matchfile(char *filestart) {
+	wordexp_t exp;
+	int i, j, k, p=strlen(filestart);
+	filestart[ p+1 ] = 0;
+	filestart[ p ] = '*';
+
+	wordexp(filestart, &exp, 0);
+	if( exp.we_wordc > 0 ) {
+		for(j=0,i=0; exp.we_wordv[0][i]!=0; i++,j++) {
+			if( exp.we_wordv[0][i]==' ' ) filestart[j++]='\\';
+			filestart[j]=exp.we_wordv[0][i];
+		}
+		filestart[j]=0;
+
+		for(k=1; k<exp.we_wordc; k++)  // comment this block for first-completion
+			for(j=0, i=0; exp.we_wordv[k][i]; i++,j++) {
+				if( filestart[j]=='\\' ) j++;
+				if( filestart[j]!=exp.we_wordv[k][i] ) {
+					filestart[j]=0;
+					break;
+				}
+			}
+	} else {
+		filestart[ p ] = 0;
+	}
+	wordfree(&exp);
+}
+
+void
 readstdin(void) {
 	char buf[sizeof text], *p, *maxstr = NULL;
 	size_t i, max = 0, size = 0;
